modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/context/ChartStateContext.js

```diff
@@ -188,19 +142,22 @@ function reducer(state, action) {
 		case CHANGE_TIME_SPAN_KEY:
 			nextState = {
 				...state,
+				loading: true,
 				timeSpanKey: action.payload.key,
 				timeSpanOffset: 0,
 			};
 			break;
 		case NEXT_TIME_SPAN:
 			nextState = {
 				...state,
+				loading: true,
 				timeSpanOffset: state.timeSpanOffset - 1,
 			};
 			break;
 		case PREV_TIME_SPAN:
 			nextState = {
 				...state,
+				loading: true,
```

Comment:
Commenting on this line only because GitHub won't let me comment on line 168 below. I gather that `default` case is never hit, right? Because `state = nextState` seems wrong (it would set `state` to `undefined` and all hell would break loose); looks like it should be `nextState = state`.

About your loading spinner bug, I can't see anything obvious that would explain it but I do have a question:

> Why is it ok to set `loading: true` directly in some of these actions but you have to go through `setLoadingState` for the `SET_LOADING` action? My gut instinct here is that you are probably firing more than one action at a time (always a potential problem, as you end up rendering at least twice) either intentionally (I can see a couple of those explicitly in there) or by accident (eg. as a result of a `useEffect`), and I suspect you are probably rendering once with `loading: true` and then immediately after with `loading: false`, which undoes the effect of the first update.

And a tip for debugging:

> I'd suggest `console.log()`-ing every action/payload and the previous/next state in the `reducer` here, along with logging the values (especially the `loading` value) you read out of the context from the component. Basically, you want to confirm that your component is rerendering when you expect it to, and that the values you see at that time are what you expect.

Logging might confirm the "gut instinct" I mentioned above. If it's true, then you might be able to find a way to tweak things to not overwrite the loading state (I see you only ever set `loading: false` in `addDataSetItem`). But even if you can, a good refactoring might be to make a couple of "composite" actions that allow you to avoid those unwanted two-in-a-row `dispatch` calls.

Reply:
Hey @wincent, thank you so much for your suggestions. I discover that I'm adding in the dependencies array `timeSpanKey` so the `useEffect` inside `Chart` is called when `timeSpanKey` changes (from 7 to 30 days and viceversa): https://github.com/liferay-frontend/liferay-portal/blob/48ca10c0e1a13e574198ac7cb54bf5368514c5f1/modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js#L207

About the two-in-a-row `dispatch` calls, you mean to keep the reduce state and then when the reduce finish, set `nextState`? (for example... 🤔 )

```js
... = reduceThings();

nextState = {...reduceState, loading: false};
```

Reply:
Oh, now I understand what you mean with two-in-a-row `dispatch` calls. I have three store with different attributes, one more related with the chart and one with the warning, if the content was published today or not, etc., and the third one for the AC connection (`validAnalyticsConnection`). You think it's a good idea to unified those different stores and keep only one with all the actions? 🤔 

I'm going to update the pull request with the loading fixed and maybe I will move this conversation into another refactor, not related to this issue. 

Thanks!

Reply:
For example, when a user changes the timeSpanKey (7-30 days), it dispatches `CHANGE_TIME_SPAN_KEY`, this will call the backend and when the promise is fulfilled, it will call `SET_METRICS`, this will change the historical data (views and/or reads) and it will trigger the `useEffect` from `Chart` component that calls `ADD_DATA_SET_ITEMS` even if there is no valid connection with AC.

Reply:
> You think it's a good idea to unified those different stores and keep only one with all the actions? 🤔

No, I think that's something for you to evaluate. It's like, you're going to eat a chocolate cake, and you get to choose how many slices to cut it into; but the amount of cake is still going to be the same. Do whatever seems "cleanest" to you.

The thing about the multiple dispatches is specifically about calling `dispatch()` two times in a row. For example, there is a `useEffect()` in `Chart.js` where you are calling `dispatch()` twice; first:

```js
dispatch({type: 'SET_LOADING'});
```

then a few lines further down:

```js
dispatch({
	// ...
	type: 'ADD_DATA_SET_ITEMS',
	// ...
});
```

So, that's going to run the reducer twice in a row, and [could lead to a double render](https://medium.com/unsplash/react-redux-performance-considerations-when-dispatching-multiple-actions-5162047bf8a6). At best it's a "smell", and at worst it becomes a user-visible flash.

Reply:
What do you think about removing `dispatch({type: 'SET_LOADING'})` and call that method inside `ADD_DATA_SET_ITEMS` dispatch? 🤔  Something like this in the store:

```js
case ADD_DATA_SET_ITEMS:
    nextState = setLoadingState(state);
    nextState = [...action.payload.keys].reduce((state, key) => {
        const dataSetItem =
            action.payload.dataSetItems?.[key] ??
            FALLBACK_DATA_SET_ITEM;

        return addDataSetItem(
            state,
            {
                ...action.payload,
                dataSetItem,
                key,
            },
            action.validAnalyticsConnection
        );
    }, state);
```

Also, I read your article and I run the Performance in the panel, I saw three times that `React Tree Reconciliation: Completed Root`. It means that I'm rendering three times the app? 😱  OMG...

Reply:
> What do you think about removing `dispatch({type: 'SET_LOADING'})` and call that method inside `ADD_DATA_SET_ITEMS` dispatch? 🤔

Seems like a definite improvement. One `dispatch`, one render, [one vision](https://www.youtube.com/watch?v=-OGd4gplxQM).